python/packages/nisar/workflows/stage_watermask.py

```diff
@@ -366,19 +382,99 @@ def download_watermask(polys, epsgs, outfile, version):
             xmin, ymin, xmax, ymax = poly.bounds
             translate_watermask(vrt_filename, outpath, xmin, xmax, ymin, ymax)
 
-        # Build vrt with downloaded watermasks
-        gdal.BuildVRT(outfile, watermask_list)
+        # Get the WATER description from the README.txt file using GDAL
+        in_readme_path = vrt_filename.replace(f'EPSG{epsg}.vrt', f'EPSG{epsg}/LICENSE.txt')
+        try:
+            water_descr = parse_readme_fields(in_readme_path)
+
+            # Build vrt with downloaded watermasks
+            vrt_dataset = gdal.BuildVRT(outfile, watermask_list)
+            # vrt_dataset.SetMetadataItem("dem_description", f'{dem_descr}')
+            for key, val in water_descr.items():
+                # # convert the bullet header to a tidy metadata key
+                # clean_key = key.lower().replace(" ", "_")
+                # vrt_dataset.SetMetadataItem(clean_key, val)
+                vrt_dataset.SetMetadataItem(key, val)
+        except Exception as e:
+            warnings.warn(
+                f"Could not read or parse README.txt at {in_readme_path} "
+                f"({e}). Proceeding without metadata."
+            )
+
     except Exception:
         errmsg = f'Failed to download NISAR WATERMASK {version} from s3 bucket. ' \
                  f'Maybe {version} is not currently supported.'
         raise ValueError(errmsg)
 
 
+def parse_readme_fields(
+        readme_path: str,
+        encoding: str = "utf-8"
+        ) -> dict[str, str]:
+    """
+    Fetch a README.txt living in an S3 bucket via GDAL and return all
+    bullet style fields in a dictionary.
+
+    Parameters
+    ----------
+    readme_path : str
+        `/vsis3/.../README.txt` or any other GDAL compatible URI.
+    encoding : str, default "utf-8"
+        Text encoding to decode the bytes.
+
+    Returns
+    -------
+    fields : dict[str, str]
+        Mapping from the bullet header (e.g. "Short description")
+        to its full text (potentially multi line).
+    """
+    # ---- 1. Pull the file into memory ----------------------------------
+    fp = gdal.VSIFOpenL(readme_path, "rb")
+    if fp is None:
+        raise IOError(f"Could not open {readme_path} via GDAL")
+
+    try:
+        data = gdal.VSIFReadL(1, 10_000_000, fp)      # read up to 10 MB
+    finally:
+        gdal.VSIFCloseL(fp)
+
+    text = data.decode(encoding, errors="replace")
+
+    # ---- 2. Parse bullet fields ----------------------------------------
+    fields = {}
+    curr_key, curr_val_lines = None, []
+
+    for line in text.splitlines():
+        bullet_match = re.match(r"^\s*-\s+(.*?):\s*(.*)$", line)
```

Comment:
This is recompiling the regex in each iteration of the `for` loop. It'd be more efficient to compile the regex before entering the loop.

```suggestion
    regex = re.compile(r"^\s*-\s+(.*?):\s*(.*)$")
    for line in text.splitlines():
        bullet_match = regex.match(line)
```